pkg/controller/master/virtualmachine/vm_volume_status_controller.go

```diff
@@ -0,0 +1,493 @@
+func (h *VMVolumeStatusController) OnPVCRemove(key string, pvc *corev1.PersistentVolumeClaim) (*corev1.PersistentVolumeClaim, error) {
+	if pvc == nil {
+		return nil, nil
+	}
+
+	var refOwnerVms []string
+	var err error
+	if refOwnerVms, err = h.getPVCRefOwnerVMs(pvc); err != nil {
+		return nil, err
+	} else if len(refOwnerVms) == 0 {
+		// pvc is not attched to any owner yet
+		return nil, nil
+	}
+
+	// remove pvc status from vm
+	for _, owner := range refOwnerVms {
+		namespace, vmName := ref.Parse(owner)
+		var err error
+		var vm *kubevirtapis.VirtualMachine
+		if vm, err = h.getVM(namespace, vmName); err != nil {
+			if !errors.IsNotFound(err) {
+				logrus.Infof("fail to get owner vm:(%s/%s), give up to remove pvc:%s status, error:%s", namespace, vmName, pvc.Name, fmt.Errorf("%w", err))
+			}
+			continue
+		}
+
+		_, err = h.removeVolumeStatusFromVM(vm, volumeStatusAnnotation{Name: pvc.Name})
+		if err != nil {
+			if !errors.IsConflict(err) {
+				logrus.Infof("pvc:(%s/%s) remove status from owner vm:%s fail, error:%s", pvc.Namespace, pvc.Name, vmName, fmt.Errorf("%w", err))
+			}
+			return nil, err
+		}
+	}
+
+	return nil, nil
+}
+
+// after pvc is detached/unmounted from VM, remove its volume status from VM
+func (h *VMVolumeStatusController) removeVolumeStatusFromDetachedVM(pvc *corev1.PersistentVolumeClaim) (*corev1.PersistentVolumeClaim, error) {
+	var namespace, vmName string
+	var ok bool
+	if namespace, vmName, ok = h.decodeAnnotationDetachedVMString(pvc); !ok {
+		return pvc, nil
+	}
+
+	logrus.Debugf("remove pvc:%s status from previous owner vm:(%s/%s)", pvc.Name, namespace, vmName)
+
+	var err error
+	var vm *kubevirtapis.VirtualMachine
+	if vm, err = h.getVM(namespace, vmName); err != nil {
+		if !errors.IsNotFound(err) {
+			logrus.Infof("fail to get vm:(%s/%s), give up to remove pvc:%s status, error:%s", namespace, vmName, pvc.Name, fmt.Errorf("%w", err))
+		}
+		return pvc, nil
+	}
+
+	if _, err = h.removeVolumeStatusFromVM(vm, volumeStatusAnnotation{Name: pvc.Name}); err != nil {
+		if !errors.IsConflict(err) {
+			logrus.Infof("pvc:(%s/%s) remove status from owner vm:%s fail, error:%s", pvc.Namespace, pvc.Name, vmName, fmt.Errorf("%w", err))
+		}
+		return pvc, err
+	}
+
+	// delete this temp annotation
+	return h.removePVCAnnotationsDetachedVM(pvc)
+}
+
+func (h *VMVolumeStatusController) getPVC(namespace string, name string) (*corev1.PersistentVolumeClaim, error) {
+	if pvc, err := h.pvcCache.Get(namespace, name); err != nil {
+		if !errors.IsNotFound(err) {
+			return nil, err
+		} else {
+			return h.pvcClient.Get(namespace, name, metav1.GetOptions{})
+		}
+	} else {
+		return pvc, nil
+	}
+}
+
+func (h *VMVolumeStatusController) getVM(namespace string, name string) (*kubevirtapis.VirtualMachine, error) {
+	if vm, err := h.vmCache.Get(namespace, name); err != nil {
+		if !errors.IsNotFound(err) {
+			return nil, err
+		} else {
+			return h.vmClient.Get(namespace, name, metav1.GetOptions{})
```

Comment:
same reason as getPVC, the reverse updating will only try once and not return error when can not find upstream object, because the upstream object may have already bee deleted.